src/rust/pipeline-ingress/src/main.rs

```diff
@@ -0,0 +1,209 @@
+    Tracing(#[from] TraceError),
+}
+
+#[tracing::instrument(err)]
+async fn handler() -> Result<(), ConfigurationError> {
+    let socket_address = std::env::var("PIPELINE_INGRESS_BIND_ADDRESS")?;
+
+    let healthcheck_polling_interval_ms =
+        std::env::var("PIPELINE_INGRESS_HEALTHCHECK_POLLING_INTERVAL_MS")?.parse()?;
+
+    let bootstrap_servers = std::env::var("KAFKA_BOOTSTRAP_SERVERS")?;
+    let sasl_username = std::env::var("KAFKA_SASL_USERNAME")?;
+    let sasl_password = std::env::var("KAFKA_SASL_PASSWORD")?;
+
+    let raw_logs_topic = "raw-logs".to_string();
+
+    tracing::info!(
+        message = "configuring kafka producer",
+        bootstrap_servers = %bootstrap_servers,
+        topic = %raw_logs_topic,
+    );
+    let producer: Producer<Envelope<RawLog>> = Producer::new(
+        bootstrap_servers.clone(),
+        sasl_username,
+        sasl_password,
+        raw_logs_topic.clone(),
+    )?;
+    tracing::info!(
+        message = "kafka producer configured successfully",
+        bootstrap_servers = %bootstrap_servers,
+        topic = %raw_logs_topic,
+    );
+
+    tracing::info!(
+        message = "configuring gRPC server",
+        socket_address = %socket_address,
+    );
+    let (server, _shutdown_tx) = PipelineIngressServer::new(
+        IngressApi::new(producer),
+        TcpListener::bind(socket_address.clone()).await?,
+        || async { Ok(HealthcheckStatus::Serving) }, // FIXME: this is garbage
+        Duration::from_millis(healthcheck_polling_interval_ms),
+    );
+    tracing::info!(
+        message = "gRPC server configured successfully",
+        socket_address = %socket_address,
+        service_name = %server.service_name(),
+    );
+
+    tracing::info!(
+        message = "starting gRPC server",
+        socket_address = %socket_address,
+    );
+    Ok(server.serve().await?)
+}
+
+#[tokio::main]
+async fn main() -> Result<(), ConfigurationError> {
+    let (non_blocking, _guard) = tracing_appender::non_blocking(std::io::stdout());
+
+    // initialize json logging layer
+    let log_layer = tracing_subscriber::fmt::layer()
+        .json()
+        .with_writer(non_blocking);
+
+    // initialize tracing layer
+    global::set_text_map_propagator(TraceContextPropagator::new());
+    let tracer = opentelemetry_jaeger::new_pipeline()
+        .with_service_name("pipeline-ingress")
+        .install_batch(opentelemetry::runtime::Tokio)?;
+
+    // register a subscriber
+    let filter = EnvFilter::from_default_env();
+    tracing_subscriber::registry()
+        .with(filter)
+        .with(log_layer)
+        .with(tracing_opentelemetry::layer().with_tracer(tracer))
+        .init();
+
+    tracing::info!("logger configured successfully");
+    tracing::info!("starting up!");
+
+    match handler().await {
+        Ok(res) => {
+            tracing::info!("shutting down");
+            Ok(res)
+        }
+        Err(err) => {
+            tracing::error!("configuration error: {}", err);
```

Comment:
I'd suggest:
```
tracing::error!(message="configuration error", err=%err);
```
but it's not a big deal here

Reply:
fixed